### ISSUE TITLE
Home Position/Zoom & Other Minor Items

### ISSUE THREAD

**Issue by jjakucyk** [openseadragon/openseadragon#1403]
Please forgive my ignorance here but I'm not a programmer and know nothing about javascript and CSS, and I only know basic HTML for the most part.  

I have an extensive map on my [transit history website](http://www.jjakucyk.com/transit/) that I had built using [Bramus' Google Maps Tile Cutter script for Photoshop.](https://github.com/bramus/photoshop-google-maps-tile-cutter)  For whatever reason it went totally pear shaped yesterday, I assume due to some change in the Google Maps API.  I've also been looking for a new solution that works better on high resolution (retina) screens, since the Bramus script seems to hard code everything at 72dpi, at least as best as I can tell.  So that's why I'm here.  

I actually managed to get the map functioning, which can be seen [here.](http://www.jjakucyk.com/transit/map)  It uses Zoomify source images, which I know isn't ideal, but that's the only way I've been able to generate the tiles that doesn't cost big money or require tools I can't figure out how to use.  Yes there's a lot of software out there, and while I can handle some command line tools, if all I get is source code then that's beyond my reach.  It also needs to be Mac compatible, and it needs to support cutting an image that's 60,000 x 60,000 pixels minimum.  

Anyway, there's just a couple niggling items that I'd like to address if at all possible. 

First, I want to set the home view to be zoomed in on Cincinnati rather than the full extent of the map.  [Something like this.](http://www.jjakucyk.com/maphome.jpg)  I've not yet been able to make that work.

Second, panning and zooming is jerky in Safari, though it's fine in both Firefox and Chrome.  Could it have something to do with the number of tiles?  When I used the Bramus script I changed the tile size to 1024 instead of the default 256, but I don't see a way to change that in the Zoomify app.  

Lastly, and this is minor, but what would it take to get retina-compatible navigation buttons?  Even if I drew new assets myself, how would they be properly displayed?  

I tried to post the code for the whole page but it got mangled, so just view the page source.  Note that I coded the image size as 120,000 x 120,000 instead of the actual 60,000 x 60,000 to allow the viewer to zoom in to a reasonable size on a mobile device.  If it use 60,000 it's just too small.  I may try to change my master image to 120,000 pixels, but as it is I have to export a 4-page PDF and manually import and assemble them in Photoshop since it can only rasterize a PDF at 32,000 pixels maximum.  I'm open to other possible workflows, but haven't found an alternative yet.  

Thanks for any advice or assistance you can offer!

**Comment by iangilman** [comment]
@jjakucyk Welcome! 

Unfortunately remapping the home button to a different location isn't exactly supported yet. You can see a discussion about that here: https://github.com/openseadragon/openseadragon/issues/643

That issue includes one possible short-term fix (overwrite the `goHome` function). There are other options as well:

* If you just want to change the initial location (but you're fine with the home button zooming you all the way out), you can call `fitBounds` in the `open` event handler. 
* You can use the `defaultZoomLevel` option to change the home zoom level, zooming you in both on startup and when you hit the home button. This doesn't allow you to change the home pan, however; it'll always be the center.

If you do want it to affect the home button and you don't want it to be in the center, then you'll have to hack `goHome` as mentioned in https://github.com/openseadragon/openseadragon/issues/643 or help us actually add that feature to OSD.

Regarding your other questions: 

* Zoomify is actually just fine (though, yes, I think it's limited to just 256px tiles).
* Sorry to hear panning isn't smooth on Safari... I don't think that should be affected by the number of tiles. What computer is this on? I tried your site in Safari on my Mac and it seems fine. If this is on an iPad, it might be this issue: https://github.com/openseadragon/openseadragon/issues/678 Either way, you might try the hack suggested there. 
* To change the default buttons, just provide a different set of button images. You can point to your new set using the `prefixUrl` param when you create the viewer. I actually don't know for sure if it'll accept retina images... It might just make the buttons twice as big. If so, that would be another good enhancement to have.
* If you want to be able to zoom in further, you don't need to change the image size... Just use the `maxZoomPixelRatio` param when you create the viewer. If you want to be able to zoom in to twice real size, use `2`.

**Comment by jjakucyk** [comment]
I'm fine with the home button going to the full view, I'm more interested in setting the initial location/zoom.  Unfortunately I have no idea how to do that.  Calling `fitBounds` in the `open` event handler means basically nothing to me.  I'd need to see exactly the right kind of example code to even begin to figure out how to implement it, and then how do I get the coordinates?  

What's strange about the Safari situation is that it's smooth as butter when the page first opens, and the first time you pan around.  Once you release the mouse button and then pan or zoom again the frame rate is garbage.  This is on a brand new iMac Pro.  

I tried swapping the home_rest.png button with a different image to test, and you're right, it just displays it much bigger, regardless of the resolution set in the file.  Could the buttons be forced to a fixed size on the page, similar to the html `width="35" height="35"` but if the actual image is 70x70 it will render it properly at 144dpi?  

I figured out the `maxZoomPixelRatio` after the usual trial and error, thanks.

**Comment by iangilman** [comment]
Okay, I'm seeing the slowdown too. Odd. One thing I've noticed is a bunch of these errors in the console:

![image](https://user-images.githubusercontent.com/338989/36553826-35d2079c-17b2-11e8-9735-c09c18cb5e32.png)

Does that mean anything to you? I see you've got some tagjs code on the page... Maybe it's related?

To do the fitBounds on open, it would look like this:

```
viewer.addHandler('open', function() {
  var bounds = OpenSeadragon.Rect(0.25, 0.25, 0.5, 0.5);
  viewer.viewport.fitBounds(bounds, true);
});
```

To find exactly the rectangle you want (instead of the numbers I gave there), zoom to that location in the viewer and open up the developer console and enter `viewer.viewport.getBounds()` and it'll tell you where you are.

Regarding forcing the buttons to 35px, this should work:

```
viewer.buttons.buttons.forEach(function(button) {
    button.element.style.width = '35px';
    button.element.style.height = '35px';
    button.element.querySelectorAll('img').forEach(function(img) {
        img.style.width = '100%';
        img.style.height = '100%';
    });
});
```

**Comment by iangilman** [comment]
I've created a new issue for actually supporting high res buttons: #1404. Hopefully someone will volunteer to take it on.

**Comment by jjakucyk** [comment]
No idea about the extra javascript code.  I just uploaded a totally blank HTML file to my web server and that got injected.  I opened up a support ticket with the hosting company.  

I got some new buttons and have that all working as expected.  Yay.  I can't get the fitBounds code to do anything though.  What am I doing wrong?  

var viewer = OpenSeadragon({
            debugMode: false,
            id: "contentDiv",
			maxZoomPixelRatio:3,
            prefixUrl: "/openseadragon/images/",
            tileSources: [{
                type: "zoomifytileservice",
                width: 60000,
                height: 60000,
                tilesUrl: "/zoomify/"
		    }],
  
        gestureSettingsMouse: {
    	flickEnabled: true,
    	clickToZoom: false,
    	dblClickToZoom: true
  		},
		  zoomPerSecond: 0.5,
		  zoomPerScroll: 1.5
        });
		
		viewer.addHandler('open', function() {
			 var bounds = OpenSeadragon.Rect(0.25, 0.25, 0.5, 0.5);
			 viewer.viewport.fitBounds(bounds, true);
		});	
		
		viewer.buttons.buttons.forEach(function(button) {
    		button.element.style.width = '35px';
   			button.element.style.height = '35px';
   			button.element.querySelectorAll('img').forEach(function(img) {
        		img.style.width = '100%';
        		img.style.height = '100%';
   			});
		});	
		
		OpenSeadragon.supportsFullScreen = false;
		viewer.setFullPage(true);
		OpenSeadragon.supportsFullScreen = true;

**Comment by iangilman** [comment]
My bad... I forgot the `new`. It should be: 

```
var bounds = new OpenSeadragon.Rect(0.25, 0.25, 0.5, 0.5);
```

**Comment by jjakucyk** [comment]
Boom, that did the trick, thanks!  http://www.jjakucyk.com/transit/map/index.html 

I did some more exploration into the jerkiness, and it's definitely a Safari-specific thing (whether it's also Webkit in general, I don't know).  I did finally figure out it's not limited just to my map, but it's more noticeable there since I used a larger full page viewer window.  

When looking at the Open Seadragon front page Duomo image, it's fine in the small viewer, but make it full screen and after that first pan it becomes a total dumpster fire.  I see major lag on my GPU graph, indicating that the CPU is waiting a significant amount of time for data from the GPU.  The CPU then spikes as well, whereas during the first pan CPU usage was pretty marginal.  In Firefox and Chrome the CPU usage is minimal and there's no waiting for the GPU, whether during the first pan or subsequent ones. 

So it's almost as if Safari is not calling the GPU after the first panning operation, or otherwise throttling it.  HOWEVER, this page pans and zooms just fine in Safari, even at full screen, so what's different?  http://dump.ventero.de/xkcd1110/
